### PR TITLE
Support disambiguating links with type signature information

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://112224233

## Summary

Adds support for disambiguating symbol links with information from the symbol graph's `functionSignature` data.

This new syntax was pitched on  the Swift Forums [in this thread](https://forums.swift.org/t/human-readable-alternative-for-docc-symbol-link-disambiguation/65792).

With these changes DocC will support type signature disambiguation in links and will use it in diagnostics but topic references and page URLs will continue to only use kind-and-hash disambiguation to ensure that all pages have the same URLs as without these changes.

## Dependencies

None.

## Testing

_Describe how a reviewer can test the functionality of your PR. Provide test content to test with if
applicable._

Steps:
1. _Provide setup instructions._
2. _Explain in detail how the functionality can be tested._

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [ ] Updated documentation if necessary

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
I feel that the code in this PR is getting too long to be easily reviewed but splitting it into multiple files _in the same PR_ makes it much much harder to know what changed and what moved. Because of this I opened https://github.com/apple/swift-docc/pull/662 to only split the code into multiple files and then this PR can focus on just the new feature.

**Comment by d-ronnqvist** [comment]
I'm breaking this up into smaller pieces (#824 and #828) so that there won't be as many merge conflicts to deal with. I'll rebase this PR once both those have been merged.

**Comment by d-ronnqvist** [comment]
The main implementation is ready for review. I will update the user facing documentation to describe this new syntax.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
I resolved the merge conflicts so this is reviewable again.

I also updated the code to make some improvements for overload groups and C++ operators which are new changes since last time I updated this PR.

**Comment by d-ronnqvist** [comment]
I have some more improvements in mind for the richness of the error messages about ambiguous links, but my plan is to defer that to a follow up PR, sometime after this is merged 🤞

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    mutating func takeAll() -> Substring {
+        defer { remaining.removeAll() }
+        return remaining
+    }
+    
+    mutating func scan(until predicate: (Character) -> Bool) -> Substring? {
+        guard let index = remaining.firstIndex(where: predicate) else {
+            return nil
+        }
+        defer { remaining = remaining[index...] }
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift:258]: Are these two lines equivalent to calling `take(index)` ? Could you call that instead?

File: Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift [hunk 2]
```diff
@@ -1289,19 +1354,376 @@ class PathHierarchyTests: XCTestCase {
...
+                       "/ShapeKit/OverloadedEnum/firstTestMemberName(_:)-(String)")
         XCTAssertEqual(paths["s:8ShapeKit14OverloadedEnumO19firstTestMemberNameySdSaySdGF"],
-                       "/ShapeKit/OverloadedEnum/firstTestMemberName(_:)-88rbf")
+                       "/ShapeKit/OverloadedEnum/firstTestMemberName(_:)-([Double])")
+        XCTAssertEqual(paths["s:8ShapeKit14OverloadedEnumO19firstTestMemberNameyS2dF"],
+                       "/ShapeKit/OverloadedEnum/firstTestMemberName(_:)-(Double)")
     }
-
+    
+    func testApplyingSyntaxSugarToTypeName() {
```
Comment (patshaughnessy) [inline Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift:1370]: Good to see all these tests - this logic is quite complex.

**Review by patshaughnessy — APPROVED** [review]
Amazing work!

I left a lot of questions and a few suggestions mostly around the readability of the code. You added extensive new unit tests, which is great, but I can imagine some quirks or odd corner conditions could come up someday. I'd suggest we merge this and then keep an eye on how the disambiguation is working over time as we use these new disambiguation styles more and more.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Dump.swift [hunk 3]
```diff
@@ -21,20 +23,25 @@ private extension PathHierarchy.Node {
     /// Maps the path hierarchy subtree into a representation that can be printed into a visual form for debugging.
     func dumpableNode() -> DumpableNode {
         // Each node is printed as 3-layer hierarchy with the child names, their kind disambiguation, and their hash disambiguation.
+        
+        // One layer for the node itself that displays information about the symbol
         return DumpableNode(
-            name: symbol.map { "{ \($0.identifier.precise) : \($0.kind.identifier.identifier) [\(languages.map(\.name).joined(separator: ", "))] }" } ?? "[ \(name) ]",
-            children: children.sorted(by: \.key).map { (key, disambiguationTree) -> DumpableNode in
+            name: symbol.map(describe(_:)) ?? "[ \(name) ]",
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Dump.swift:29]: Yes, that's basically the same, or at least the compiler should be able to make them the same. 

Here `describe(_:)` is referring to [private func describe(_ symbol: SymbolGraph.Symbol) -> String][1] (on line 54 below) by its name. The underscore in the name just because the parameter doesn't have an external argument label.



[1]: https://github.com/swiftlang/swift-docc/pull/643/files#diff-efd1b7cf023713dbd75be874c85bbc4cce9f1f563b748548a39cf68a0bd037c2R54

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift [hunk 4]
```diff
@@ -243,6 +246,65 @@ private struct PathComponentScanner {
...
+        // If the next component is a C++ operator, parse the full operator before splitting on "/" ("/" may appear in the operator name)
+        if remaining.starts(with: Self.cxxOperatorPrefix),
+           remaining.unicodeScalars.dropFirst(Self.cxxOperatorPrefixLength).first?.isValidCxxOperatorSymbol == true
+        {
+            let base = scan(length: Self.cxxOperatorPrefixLength + 1)
+            // Because C++ operators don't include the parameters in the name,
+            // a trailing "-" could either be part of the name of be the disambiguation separator.
+            //
+            // The only valid C++ operators that include a "-" do so at the start.
+            // However, "-=", "->", and "->*" don't have a trailing "", so they're unambiguous.
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift:264]: Yes, thanks for catching this. It should be a `-`

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 5]
```diff
@@ -182,7 +190,24 @@ extension PathHierarchy.DisambiguationContainer {
...
     static func disambiguatedValues(
         for elements: some Sequence<Element>,
-        includeLanguage: Bool = false
+        includeLanguage: Bool = false,
+        allowAdvancedDisambiguation: Bool = true
+    ) -> [(value: PathHierarchy.Node, disambiguation: Disambiguation)] {
+        var collisions = _disambiguatedValues(for: elements, includeLanguage: includeLanguage, allowAdvancedDisambiguation: allowAdvancedDisambiguation)
+        
+        // If all but one of the collisions are disfavored, remove the disambiguation for the only favored element.
+        if let onlyFavoredElementIndex = collisions.firstIndex(where: { !$0.value.specialBehaviors.contains(.disfavorInLinkCollision) }),
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift:199]: Oh yeah, I'm accessing the option set directly. I can create convenience accessors for these.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Error.swift [hunk 6]
```diff
@@ -80,10 +81,38 @@ extension PathHierarchy.Error {
...
+                .map { (fullName: fullNameOfNode($0.node), disambiguation: $0.disambiguation) }
+                .sorted { lhs, rhs in
+                    // Sort by name first and disambiguation second
+                    if lhs.fullName == rhs.fullName {
+                        return lhs.disambiguation < rhs.disambiguation
+                    }
+                    return lhs.fullName < rhs.fullName
+                }
+                .map { (fullName: String, suggestedDisambiguation: String) -> Solution in
+                    // In contexts that display the solution message on a single line by removing newlines, this extra whitespace makes it look correct ─────────────╮
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Error.swift:109]: I am not sure I understand what this is referring to.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Find.swift [hunk 7]
```diff
@@ -558,3 +578,11 @@ private extension PathHierarchy.Node {
             || keys.contains(String(component.name))
     }
 }
+
+private func typesMatch(provided: [Substring], actual: [String]?) -> Bool {
+    guard let actual = actual, provided.count == actual.count else { return false }
+    for (providedType, actualType) in zip(provided, actual) where providedType != "_" && providedType != actualType {
```
[the remark this answers is not in the document]
Comment (daniel-grumberg) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Find.swift:584]: I for one prefer the `allSatisfy` version as it's a bit more explicit, I assume that `allSatisfy` implements the short circuiting-behavior as well.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift [hunk 8]
```diff
@@ -243,6 +246,65 @@ private struct PathComponentScanner {
         return scanUntilSeparatorAndThenSkipIt()
     }
     
+    mutating func _scanOperatorName() -> Substring? {
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift:249]: Should this be private?

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+            // We don't want to list "void" return values as type disambiguation
+            return nil
+        }
+        return typeSpellings(for: fragments)
+    }
+    
+    private static let knownVoidReturnValues = ParametersAndReturnValidator.knownVoidReturnValuesByLanguage.flatMap { $0.value }
+    
+    private static func typeSpellings(for fragments: [SymbolGraph.Symbol.DeclarationFragments.Fragment]) -> String? {
+        var accumulated = ""[...]
```
[the remark this answers is not in the document]
Comment (daniel-grumberg) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift:44]: yeah although I don't see why that is needed here.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Error.swift [hunk 10]
```diff
@@ -80,10 +81,38 @@ extension PathHierarchy.Error {
...
+        // Both `.unknownDisambiguation(...)` and `.lookupCollisions(...)` create solutions on the same format from the same information.
         // This is defined inline because it captures `fullNameOfNode`.
-        func collisionIsBefore(_ lhs: (node: PathHierarchy.Node, disambiguation: String), _ rhs: (node: PathHierarchy.Node, disambiguation: String)) -> Bool {
-            return fullNameOfNode(lhs.node) + lhs.disambiguation
-                 < fullNameOfNode(rhs.node) + rhs.disambiguation
+        func makeCollisionSolutions(
+            from candidates: [(node: PathHierarchy.Node, disambiguation: String)],
+            nextPathComponent: PathHierarchy.PathComponent,
+            partialResultPrefix: Substring
+        ) -> (
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Error.swift:90]: Possibly. It's internal to this function so only this function can call it.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Find.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Find.swift:584]: Either works for me.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+        for fragment in fragments {
+            switch fragment.kind {
+            case .identifier where fragment.preciseIdentifier == nil, // Skip the argument label
+                 .keyword,   // Skip keywords ("inout", "consuming", "each", etc.)
+                 .attribute, // Skip attributes ("@escaping", custom result builders, etc.)
+                 .numberLiteral, .stringLiteral, // Skip literals
+                 .externalParameter, .genericParameter, .internalParameter:
+                continue
+
+            default:
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift:55]: Should we be more explicit and only capture the fragment kinds we care about, that would make me feel a bit better if we start seeing unexpected values.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+    func peek() -> Character? {
+        remaining.first
+    }
+    
+    mutating func take() -> Character {
+        remaining.removeFirst()
+    }
+    
+    mutating func take(_ count: Int) -> Substring {
+        defer { remaining = remaining.dropFirst(count) }
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift:245]: You are correct. This mutation happens after returning the original value. 

In some cases like this I find that it's a good way of expressing a pattern that returns a value and makes modifications. If this code was longer I would probably use an intermediate value instead.

[the remark this answers is not in the document]
Comment (daniel-grumberg) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift:245]: Mega-nit: 
```suggestion
        defer { remaining.removeFirst(count) }
```

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift:199]: > Can you use the `singleMatch` extension to Sequence here?

No, because the collision elements are value types and I want to in-place modify the array of collisions.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift (hunk 1 above)
Comment (daniel-grumberg) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift:258]: Crucially, Substring (nor anything StringProtocol-y for that matter) doesn't conform to `RandomAccessCollection`, which is the protocol that makes it possible to calculate [`distance(from:to:)`](https://developer.apple.com/documentation/swift/collection/distance(from:to:)-4r6r8)) in O(1) time. The reason why is precisely because of David's explanation above.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift (hunk 8 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift:249]: It can't be. I extracted this code so that `PathParser.parseOperatorName(_:)` (a few lines above) could call it.

The entire `PathComponentScanner` type is file private, so this is also implicitly file private.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 13]
```diff
@@ -19,19 +19,25 @@ private func symbolFileName(_ symbolName: String) -> String {
...
     /// - Parameters:
     ///   - includeDisambiguationForUnambiguousChildren: Whether or not descendants unique to a single collision should maintain the containers disambiguation.
     ///   - includeLanguage: Whether or not kind disambiguation information should include the source language.
+    ///   - allowAdvancedDisambiguation: Whether or not to support more advanced and more human readable types of disambiguation.
     /// - Returns: A map of unique identifier strings to disambiguated file paths.
     func caseInsensitiveDisambiguatedPaths(
         includeDisambiguationForUnambiguousChildren: Bool = false,
-        includeLanguage: Bool = false
+        includeLanguage: Bool = false,
+        allowAdvancedDisambiguation: Bool = true
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift:33]: These values aren't user facing feature flags. They're only default values for this function. They're never stored anywhere. Instead the caller passes the flags for the behaviors that it needs to customize.

For all three parameters, these default value is what results in the least amount of disambiguation and the most readable disambiguation. 

The only places that passes custom flags for this is `PathHierarchyBasedLinkResolver.referencesForSymbols(in:bundle:context:)` which the context calls to create the symbol pages's references (which is later used to create their URLs). 

To continue to produce the same references and URLs as before, we include some unnecessary disambiguation, we include languages (even though it doesn't provide any disambiguation information), and we limit the results to kind-and-hash disambiguation.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+        //    parameter types         Result,(Result,Element)->Result     │
+        //                            ╰─┬──╯ ╰──────────┬───────────╯     │
+        //    first parameter type    Result            │                 │
+        //    second parameter type          (Result,Element)->Result     │
+        //                                                                │
+        //    return type(s)                                            Result
+        
+        let possibleDisambiguationText: Substring
+        if let name = parseOperatorName(original) {
+            print("\(name)\n\(original)\n\n")
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift]: Removed in https://github.com/swiftlang/swift-docc/pull/643/commits/30e71ada5508ba05e2cb05a014e1389350bfe8d0

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift]: Updated in https://github.com/swiftlang/swift-docc/pull/643/commits/7f7f5f5e3d73771259f650f0f48bfa72010edee4

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 15]
```diff
@@ -203,6 +228,73 @@ extension PathHierarchy.DisambiguationContainer {
             return collisions
         }
         
+        if allowAdvancedDisambiguation {
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift:229]: Extracted in https://github.com/swiftlang/swift-docc/pull/643/commits/528a45d705bd7bcc2152f3bc0b53f3dba9eb18ed

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift]: Convenience accessors added in https://github.com/swiftlang/swift-docc/pull/643/commits/216a0383e5a0a1d6a4f6cca067439a4ed28e0349

`onlyIndex(where:)` utility extracted in https://github.com/swiftlang/swift-docc/pull/643/commits/d2045a8f94fb96245a8d451d8479903c0c241acc

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift [hunk 16]
```diff
@@ -243,6 +246,65 @@ private struct PathComponentScanner {
...
+            return scanUntil(index: remaining.firstIndex(of: Self.swiftOperatorEnd)) + scan(length: 1)
+        }
+        
+        // If the next component is a C++ operator, parse the full operator before splitting on "/" ("/" may appear in the operator name)
+        if remaining.starts(with: Self.cxxOperatorPrefix),
+           remaining.unicodeScalars.dropFirst(Self.cxxOperatorPrefixLength).first?.isValidCxxOperatorSymbol == true
+        {
+            let base = scan(length: Self.cxxOperatorPrefixLength + 1)
+            // Because C++ operators don't include the parameters in the name,
+            // a trailing "-" could either be part of the name of be the disambiguation separator.
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift]: Fixed in https://github.com/swiftlang/swift-docc/pull/643/commits/eca9ce5bf73eb26e0763c512a5affb31c834749e

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift]: Fixed in https://github.com/swiftlang/swift-docc/pull/643/commits/eca9ce5bf73eb26e0763c512a5affb31c834749e

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Find.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Find.swift]: Updated to use `allSatisfy` and rephrase the condition to a positive expression instead of a negative expression in https://github.com/swiftlang/swift-docc/pull/643/commits/316351cb938f747704ce7228ae03e0c07d93dda7

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift (hunk 11 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift:53]: Updated in https://github.com/swiftlang/swift-docc/pull/643/commits/7c99db64a9bb1cfc9e6efc81fd06a29b5be31d47

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Error.swift (hunk 6 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Error.swift:109]: The default diagnostic formatter outputs diagnostic solutions (phrased as "suggestions") on a single line. For example:

```
4 | /// Here's a link that uses the wrong symbol kind disambiguation in a link:
5 + /// - ``myFunction(with:and:)-struct``
  |                              ╰─suggestion: Remove '-struct' for'func myFunction(with one: Int, and two: String) -> Bool'
```

If you look closely at the suggestion text you can see that there's a missing space between "for" and the symbol name.

With these changes this suggestion text isn't missing that space:
```
suggestion: Remove '-struct' for 'func myFunction(with one: Int, and two: String) -> Bool'
                                ^
                                with this change there is an added space here
```

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 17]
```diff
@@ -295,6 +404,14 @@ extension PathHierarchy.DisambiguationContainer {
                 return ""
             case .kind(let value), .hash(let value):
                 return "-"+value
+            case .returnTypes(let types):
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift:354]: Added in https://github.com/swiftlang/swift-docc/pull/643/commits/2e18f6b7965150b237cedec01a85791e2d6a32fe

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 18]
```diff
@@ -47,7 +53,7 @@ extension PathHierarchy {
         
         func gatherLinksFrom(_ containers: some Sequence<DisambiguationContainer>) {
             for container in containers {
-                let disambiguatedChildren = container.disambiguatedValuesWithCollapsedUniqueSymbols(includeLanguage: false)
+                let disambiguatedChildren = container.disambiguatedValuesWithCollapsedUniqueSymbols(includeLanguage: false, allowAdvancedDisambiguation: true)
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift:56]: I'm not _too_ concerned about this since these are mostly file private arguments that are passed by the caller.

**Comment by d-ronnqvist** [comment]
I started updating the user facing documentation related to link disambiguation but found that I want to make slightly larger documentation changes that I first thought and I feel that those updates would benefit from a separate review. 

I will merge this PR now since it's a big improvement and there doesn't seem to be any _major_ feedback anymore. If there's more changes that you're like to request, please message me and I'll make a follow up PR. I already have further refinements and enhancements in mind so I plan to continue to iterate on this code over the coming months anyway.
